File: backend/app/models/trip.py
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


class TripMember(BaseModel):
    user_id: str = Field(alias="userId")
    role: str = Field(default="editor", pattern="^(owner|editor|viewer)$")

    class Config:
        populate_by_name = True


class TripDates(BaseModel):
    start: Optional[datetime] = None
    end: Optional[datetime] = None


class TripBase(BaseModel):
    title: str
    destination: Optional[str] = None
    dates: Optional[TripDates] = None
    status: str = Field(default="draft", pattern="^(draft|planned|booked)$")
    readiness: int = Field(default=0, ge=0, le=100)
    cover_image: Optional[str] = None


class TripCreate(TripBase):
    pass


class TripInDB(TripBase):
    id: Optional[str] = Field(default=None, alias="_id")
    members: List[TripMember] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        populate_by_name = True


class Trip(TripBase):
    id: str
    members: List[dict]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: backend/src/models/trip.ts
import { z } from 'zod';

export const TripMemberSchema = z.object({
  userId: z.string(),
  role: z.enum(['owner', 'editor', 'viewer']).default('editor'),
});

export const TripDatesSchema = z.object({
  start: z.date().optional(),
  end: z.date().optional(),
});

export const TripBaseSchema = z.object({
  title: z.string(),
  destination: z.string().optional(),
  dates: TripDatesSchema.optional(),
  status: z.enum(['draft', 'planned', 'booked']).default('draft'),
  readiness: z.number().min(0).max(100).default(0),
  cover_image: z.string().optional(),
});

export const TripCreateSchema = TripBaseSchema;

export type TripMember = z.infer<typeof TripMemberSchema>;
export type TripDates = z.infer<typeof TripDatesSchema>;
export type TripBase = z.infer<typeof TripBaseSchema>;
export type TripCreate = z.infer<typeof TripCreateSchema>;

export interface TripInDB extends TripBase {
  _id?: string;
  tripId?: string;
  members: TripMember[];
  createdAt?: Date;
  updatedAt?: Date;
}

export interface Trip extends TripBase {
  id: string;
  members: Array<{ userId: string; role: string }>;
  created_at: Date;
  updated_at: Date;
}
